R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+#' @details aggregate individual weights and outcomes to cluster weighted sums
+#' @return a list of 
+#' - a data frame of cluster weights, outcomes, treatments, and block ids;
+#' - treatment id column name; 
+#' - block id column name
+#' @keywords internal
+.aggregate_to_cluster <- function(x, ...){
+  ws <- if (is.null(x$weights)) 1 else x$weights
+  data_temp <- x$call$data
+  name_y <- as.character(x$terms[[2]]) # the column of y
+  data_temp <- cbind(data_temp, .w = ws, .w0 = ws / ate(x@Design, data=data_temp),
+                     .wy = ws * data_temp[, name_y])
+  
+  design_obj <- x@Design
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  
+  data_temp <- .merge_block_id_cols(data_temp, name_blk)
+  name_blk <- name_blk[1]
+  
+  eq_clu <- paste(name_clu, collapse = " + ")
+  form <- paste("cbind(", name_trt, ", ", name_blk, ") ~ ", eq_clu, sep = "")
+  form2 <- paste("cbind(.wy, .w, .w0) ~ ", eq_clu, sep = "")
+  
+  # aggregate z and bid by mean, aggregate w, w0, and w*y by sum
+  data_aggr <- cbind(
+    do.call("aggregate", list(as.formula(form), FUN = mean, data = data_temp)),
+    do.call("aggregate", list(as.formula(form2), FUN = sum, data = data_temp))
+  )
+  data_aggr$.wy <- data_aggr$.wy / data_aggr$.w
+  data_aggr <- data_aggr[order(data_aggr[, name_blk]), ]
+  return(list(data = data_aggr, block = name_blk, z = name_trt))
+}
+
+#' @title (Internal) Merge multiple block IDs
+#' @param df a data frame 
+#' @param ids a vector of block IDs, column names of df
+#' @details merge multiple block ID columns by the value combinations
+#' and store the new block ID in the column \code{ids[1]}
+#' @return a data frame with a column that contains unique block number IDs
+#' @keywords internal
+.merge_block_id_cols <- function(df, ids){
+  df[,ids[1]] <- apply(df[, ids, drop = FALSE], 1, paste, collapse = "_")
+  unique_ids <- data.frame(unique(df[,ids[1]]))
+  colnames(unique_ids) <- ids[1]
+  unique_ids$.ID <- seq_len(nrow(unique_ids))
+  
+  df <- merge(df, unique_ids, by = ids[1], all.x = TRUE)
+  df[,ids[1]] <- df$.ID
+  df$.ID <- NULL
+  return(df)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @param x a fitted \code{teeMod} model
+#' @keywords internal
+.prepare_design_matrix <- function(x, ...) {
+  design_obj <- x@Design
+  data <- x$call$data
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  cid <- .merge_block_id_cols(data, name_clu)[, name_clu[1]]
+  
+  agg <- .aggregate_to_cluster(x)
+  bid <- agg[[1]][, agg[[2]]]  # block ids
+  
+  name_y <- as.character(x$terms[[2]]) # name of the outcome column
+  X1 <- model.matrix(x$call$offset@fitted_covariance_model) # design matrix
+  p <- ncol(X1)
+  
+  wc <- x$call$offset@fitted_covariance_model$weight
```

Comment:
Same thing here about call vs. model frame

Reply:
Updated in commit https://github.com/benbhansen-stats/propertee/commit/6eb8927664674088f61b3323683e7d592a2de431